### DISCUSSION TITLE
SQL syntax error when executing the query to get n random content items

### DISCUSSION THREAD

**Discussion by aminmohammadi05** [OrchardCMS/OrchardCore#12739] [Q&A] [answered]
Hi, would you please do a favor to me? I want to run this query on Orchard that return 3 random products:
`SELECT TOP (3) * FROM ContentItemIndex WHERE (ContentType = 'Product') AND (Published = 1) ORDER BY newid()`
The query works on SQL server very good but in Orchard gives this error:
`Syntax error, expected: OVER, AS, id_simple, ,, HAVING, ORDER, LIMIT, OFFSET, ), UNION, ;, SELECT, FROM, WHERE, GROUP at line:0, col:15`
![Lcmow](https://user-images.githubusercontent.com/20872120/198878774-c75a68ff-5f55-4f04-874b-228834934d7f.png)

**Comment by hishamco** [comment]
Which version of OC you are using?

**Reply by aminmohammadi05** [reply]
Orchard 1.4

**Reply by hishamco** [reply]
Could you please try the latest bits? FYI there was a change happen in the SQL Parser recently I thought it was the issue, for that please try with what I suggest and let us know

**Reply by aminmohammadi05** [reply]
I am using the nuget package already and I believe it is the latest one. do you mean that I should use OrchardCMS/OrchardCore in this repository?

**Comment by lampersky** [comment, marked as answer]
@aminmohammadi05 there are few limitations in current grammar/parser as a workaround please use sth like this:

```
SELECT newid() as randomId, cii.ContentType FROM ContentItemIndex cii
WHERE (ContentType = 'Product') AND (Published = 1) 
ORDER BY randomId limit 3
```

- TOP clause is SQL Server specific please use limit instead,
- you can't use function in order by clause,
- you can't combine * with specified column list, you need to list them all.

**Reply by aminmohammadi05** [reply]
Thanks man! you saved my life :)

**Reply by aminmohammadi05** [reply]
now another error when I try to fetch data using GraphQl: Error in parsing
![image](https://user-images.githubusercontent.com/20872120/198958900-c54823b6-a56a-420b-bec4-b16788c31acb.png)

**Reply by lampersky** [reply]
@aminmohammadi05 try to convert this **uniqueidentifier** column to varchar

```
SELECT convert(varchar(36), newid()) as randomId, cii.ContentType FROM ContentItemIndex cii
WHERE (ContentType = 'Product') AND (Published = 1) 
ORDER BY randomId limit 3
```

**Reply by aminmohammadi05** [reply]
This time another error:
![image](https://user-images.githubusercontent.com/20872120/198968038-fa24eedc-7ac6-4b8f-9760-59065963fecc.png)

**Reply by aminmohammadi05** [reply]
I think the dashes of the randomId should be removed before conversion to varchar

**Reply by lampersky** [reply]
@aminmohammadi05 try to get rid of that additional **randomId** column (by moving it to having clause which allows to use functions there), this query is a bit tricky and I would not use it in prod, but prepared it for you just for testing purpose, if this will work, then I would suggest to use latest orchard version and rewrite query to use subquery or cte

```
SELECT cii.ContentType, cii.ContentItemId
FROM ContentItemIndex cii
WHERE (ContentType = 'Product') AND (Published = 1)
GROUP by cii.ContentType, cii.ContentItemId
HAVING cii.ContentItemId > convert(varchar(36), newid())
ORDER BY ContentItemId limit 3
```

please note, you need to specify column list twice, one in select clause and then in group by clause

when you will upgrade to latest version, use this query instead
```
with test as (
	select newid() as randId, ContentItemId, DisplayText
	from ContentItemIndex
        order by randId limit 3
)
select ContentItemId, DisplayText from test 
```

**Reply by aminmohammadi05** [reply]
Again when I removed the dashes I got the same error.
The goal of this piece of code is returning 3 random products. If other approaches are there to do so I will appreciate.

**Reply by aminmohammadi05** [reply]
The error is gone but no result!!
![image](https://user-images.githubusercontent.com/20872120/198974582-f5f4b7d9-7e4b-4412-aae7-fa243058745c.png)

**Reply by aminmohammadi05** [reply]
I changed the Schema to this:
```
{
    "type": "object",
    "properties": {  
        "Name" : {
            "type" : "string",
            "description" : "This is the name of the product."
        },
      "Id" : {
            "type" : "Integer",
            "description" : "This is the Id of the product."
        },
       "ContentItemId" : {
            "type" : "string",
            "description" : "This is the ContentItemId of the product."
        }
    }
}
```
But again the same result
